crates/miden-objects/src/transaction/tx_header.rs

```diff
@@ -0,0 +1,119 @@
+use alloc::vec::Vec;
+
+use vm_processor::{DeserializationError, Digest};
+
+use crate::{
+    note::NoteId,
+    transaction::{AccountId, Nullifier, TransactionId},
+    utils::{ByteReader, ByteWriter, Deserializable, Serializable},
+};
+
+/// A transaction header derived from a
+/// [`ProvenTransaction`](crate::transaction::ProvenTransaction).
+///
+/// The header is essentially a direct copy of the transaction's commitments, in particular the
+/// initial and final account state commitment as well as all nullifiers of consumed notes and all
+/// note IDs of created notes. While account updates may be aggregated and notes may be erased as
+/// part of batch and block building, the header retains the original transaction's data.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TransactionHeader {
+    id: TransactionId,
+    account_id: AccountId,
+    initial_state_commitment: Digest,
+    final_state_commitment: Digest,
+    input_notes: Vec<Nullifier>,
+    output_notes: Vec<NoteId>,
+}
```

Comment:
One thing we cannot do with this is to verify that this struct is internally consistent. That is:

- We cannot confirm account changes and input/output notes in fact result in the provided ID.
- We cannot confirm that the listed account ID in fact belongs to the account described by initial/final commitments.

The first point can be solved by adding more data to input/output notes (e.g., note ID and metadata). But I'm not sure how to solve the second point. And if we can't solve both, not sure it makes sense to solve one of them.